File: backend-java/src/test/java/com/financepulse/engine/application/usecases/RegisterUserUseCaseTest.java
package com.financepulse.engine.application.usecases;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

import com.financepulse.engine.domain.user.Email;
import com.financepulse.engine.domain.user.User;
import com.financepulse.engine.domain.user.errors.DuplicateEmailException;
import com.financepulse.engine.domain.user.errors.WeakPasswordException;
import com.financepulse.engine.testsupport.FakePasswordHasher;
import com.financepulse.engine.testsupport.InMemoryUserRepository;
import com.financepulse.engine.testsupport.SequentialIdGenerator;
import java.util.Optional;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class RegisterUserUseCaseTest {

    private InMemoryUserRepository userRepository;
    private RegisterUserUseCase useCase;

    @BeforeEach
    void setUp() {
        userRepository = new InMemoryUserRepository();
        useCase = new RegisterUserUseCase(userRepository, new FakePasswordHasher(), new SequentialIdGenerator("user"));
    }

    @Test
    void registersANewUserAndPersistsAHashedPassword() {
        RegisterUserUseCase.Output result = useCase.execute(new RegisterUserUseCase.Input("user@example.com", "StrongPass1"));

        assertThat(result.userId()).isEqualTo("user-1");

        Optional<User> saved = userRepository.findByEmail(Email.create("user@example.com"));
        assertThat(saved).isPresent();
        assertThat(saved.get().getPasswordHash()).isEqualTo("hashed:StrongPass1");
        assertThat(saved.get().getPasswordHash()).isNotEqualTo("StrongPass1");
    }

    @Test
    void rejectsRegistrationWithAnEmailAlreadyInUseCaseInsensitively() {
        useCase.execute(new RegisterUserUseCase.Input("user@example.com", "StrongPass1"));

        assertThatThrownBy(() -> useCase.execute(new RegisterUserUseCase.Input("USER@example.com", "AnotherPass1")))
                .isInstanceOf(DuplicateEmailException.class);
    }

    @Test
    void rejectsAPasswordShorterThanTheMinimumLength() {
        assertThatThrownBy(() -> useCase.execute(new RegisterUserUseCase.Input("user@example.com", "short")))
                .isInstanceOf(WeakPasswordException.class);
    }

    @Test
    void doesNotPersistAUserWhenThePasswordIsWeak() {
        assertThatThrownBy(() -> useCase.execute(new RegisterUserUseCase.Input("user@example.com", "short")));

        Optional<User> saved = userRepository.findByEmail(Email.create("user@example.com"));
        assertThat(saved).isEmpty();
    }
}

File: backend/src/application/use-cases/RegisterUserUseCase.test.ts
import { beforeEach, describe, expect, it } from "vitest";
import { RegisterUserUseCase } from "./RegisterUserUseCase";
import { InMemoryUserRepository } from "../../adapters/out/persistence/InMemoryUserRepository";
import { Email } from "../../domain/user/Email";
import { DuplicateEmailError } from "../../domain/user/errors/DuplicateEmailError";
import { WeakPasswordError } from "../../domain/user/errors/WeakPasswordError";
import { FakePasswordHasher } from "../../test-support/FakePasswordHasher";
import { SequentialIdGenerator } from "../../test-support/SequentialIdGenerator";

describe("RegisterUserUseCase", () => {
  let userRepository: InMemoryUserRepository;
  let useCase: RegisterUserUseCase;

  beforeEach(() => {
    userRepository = new InMemoryUserRepository();
    useCase = new RegisterUserUseCase(userRepository, new FakePasswordHasher(), new SequentialIdGenerator("user"));
  });

  it("registers a new user and persists a hashed password (RF-001)", async () => {
    const result = await useCase.execute({ email: "user@example.com", password: "StrongPass1" });

    expect(result.userId).toBe("user-1");

    const saved = await userRepository.findByEmail(Email.create("user@example.com"));
    expect(saved?.passwordHash).toBe("hashed:StrongPass1");
    expect(saved?.passwordHash).not.toBe("StrongPass1");
  });

  it("rejects registration with an email already in use, case-insensitively (RF-002)", async () => {
    await useCase.execute({ email: "user@example.com", password: "StrongPass1" });

    await expect(useCase.execute({ email: "USER@example.com", password: "AnotherPass1" })).rejects.toThrow(
      DuplicateEmailError,
    );
  });

  it("rejects a password shorter than the minimum length", async () => {
    await expect(useCase.execute({ email: "user@example.com", password: "short" })).rejects.toThrow(
      WeakPasswordError,
    );
  });

  it("does not persist a user when the password is weak", async () => {
    await expect(useCase.execute({ email: "user@example.com", password: "short" })).rejects.toThrow();

    const saved = await userRepository.findByEmail(Email.create("user@example.com"));
    expect(saved).toBeNull();
  });
});
